libfreerdp/primitives/primitives.c

```diff
@@ -23,65 +24,483 @@
+}
+
+
+#if defined(WITH_OPENCL)
+static primitives_opencl_context openclContext;
+
+primitives_opencl_context *primitives_get_opencl_context(void)
+{
+	return &openclContext;
+}
+
+pstatus_t primitives_uninit_opencl(void)
+{
+	if (!openclContext.support)
+		return PRIMITIVES_SUCCESS;
+
+	clReleaseProgram(openclContext.program);
+	clReleaseCommandQueue(openclContext.commandQueue);
+	clReleaseContext(openclContext.context);
+	clReleaseDevice(openclContext.deviceId);
+	openclContext.support = FALSE;
+	return PRIMITIVES_SUCCESS;
+}
+
+BOOL primitives_init_opencl_context(primitives_opencl_context *cl)
+{
+	cl_platform_id *platform_ids = NULL;
+	cl_uint ndevices, nplatforms, i;
+	cl_kernel kernel;
+	cl_int ret;
+	char sourcePath[1000];
+	primitives_t optimized;
+
+	BOOL gotGPU = FALSE;
+	FILE *f;
+	size_t programLen;
+	char *programSource;
+
+	if (!primitives_init_optimized(&optimized))
+		return FALSE;
+	cl->YUV420ToRGB_backup = optimized.YUV420ToRGB_8u_P3AC4R;
+
+	ret = clGetPlatformIDs(0, NULL, &nplatforms);
+	if (ret != CL_SUCCESS || nplatforms < 1)
+		return FALSE;
+
+	platform_ids = calloc(nplatforms, sizeof(*platform_ids));
+	if (!platform_ids)
+		return FALSE;
+
+	ret = clGetPlatformIDs(nplatforms, platform_ids, &nplatforms);
+	if (ret != CL_SUCCESS)
+	{
+		free(platform_ids);
+		return FALSE;
+	}
+
+	for (i = 0; (i < nplatforms) && !gotGPU; i++)
+	{
+		cl_device_id device_id;
+		cl_context context;
+		char platformName[1000];
+		char deviceName[1000];
+
+		ret = clGetPlatformInfo(platform_ids[i], CL_PLATFORM_NAME, sizeof(platformName), platformName, NULL);
+		if (ret != CL_SUCCESS)
+			continue;
+
+		ret = clGetDeviceIDs(platform_ids[i], CL_DEVICE_TYPE_GPU, 1, &device_id, &ndevices);
+		if (ret != CL_SUCCESS)
+			continue;
+
+		ret = clGetDeviceInfo(device_id, CL_DEVICE_NAME, sizeof(deviceName), deviceName, NULL);
+		if (ret != CL_SUCCESS)
+		{
+			WLog_ERR(TAG, "openCL: unable get device name for platform %s", platformName);
+			clReleaseDevice(device_id);
+			continue;
+		}
+
+		context = clCreateContext(NULL, 1, &device_id, NULL, NULL, &ret);
+		if (ret != CL_SUCCESS)
+		{
+			WLog_ERR(TAG, "openCL: unable to create context for platform %s, device %s", platformName, deviceName);
+			clReleaseDevice(device_id);
+			continue;
+		}
+
+		cl->commandQueue = clCreateCommandQueue(context, device_id, 0, &ret);
+		if (ret != CL_SUCCESS)
+		{
+			WLog_ERR(TAG, "openCL: unable to create command queue");
+			clReleaseContext(context);
+			clReleaseDevice(device_id);
+			continue;
+		}
+
+		WLog_INFO(TAG, "openCL: using platform=%s device=%s", platformName, deviceName);
+
+		cl->platformId = platform_ids[i];
+		cl->deviceId = device_id;
+		cl->context = context;
+		gotGPU = TRUE;
+	}
+
+	free(platform_ids);
+
+	if (!gotGPU)
+	{
+		WLog_ERR(TAG, "openCL: no GPU found");
+		return FALSE;
+	}
+
+	snprintf(sourcePath, sizeof(sourcePath), "%s/primitives.cl", OPENCL_SOURCE_PATH);
+
+	f = fopen(sourcePath, "r");
```

Comment:
Why import that at runtime? While nice to allow changing the implementation it is a hindrance if you compile a single binary (like `wfreerdp.exe`)

Reply:
well to my knowledge there's no way to have a precompiled version as it depends on the driver, it's like a shader. The other solution would be to embed the source in the freerdp binary...

Reply:
this could, of course, be done optionally ;)